internal/archive/credentials.go

```diff
@@ -0,0 +1,299 @@
+	Password string
+}
+
+// Empty checks whether c represents unsuccessful search.
+func (c credentials) Empty() bool {
+	return c.Username == "" && c.Password == ""
+}
+
+// credentialsQuery contains parsed input URL data used for search.
+type credentialsQuery struct {
+	scheme     string
+	host       string
+	port       string
+	path       string
+	needScheme bool
+}
+
+// parseRepoURL parses repoURL into credentialsQuery and fills provided
+// credentials with username and password if they are specified in repoURL.
+func parseRepoURL(repoURL string) (creds credentials, query *credentialsQuery, err error) {
+	u, err := url.Parse(repoURL)
+	if err != nil {
+		return
+	}
+
+	creds.Username = u.User.Username()
+	creds.Password, _ = u.User.Password()
+
+	if !creds.Empty() {
+		return
+	}
+
+	host := u.Host
+	port := u.Port()
+	if port != "" {
+		// u.Hostname() would remove brackets from IPv6 address but we
+		// need it verbatim for string search in netrc file. This is
+		// also faster because both u.Port() and u.Hostname() parse
+		// u.Host into port and hostname.
+		host = u.Host[0 : len(u.Host)-len(port)-1]
+	}
+
+	query = &credentialsQuery{
+		scheme: u.Scheme,
+		host:   host,
+		port:   port,
+		path:   u.Path,
+		// If the input URL specifies unencrypted scheme, the scheme in
+		// machine declarations in netrc file is not optional and must
+		// also match.
+		needScheme: u.Scheme != "https" && u.Scheme != "tor+https",
+	}
+
+	return
+}
+
+// findCredentials searches credentials for repoURL in configuration files in
+// directory specified by CHISEL_AUTH_DIR environment variable if it's
+// non-empty, otherwise /etc/apt/auth.conf.d.
+func findCredentials(repoURL string) (credentials, error) {
+	credsDir := "/etc/apt/auth.conf.d"
+	if v := os.Getenv("CHISEL_AUTH_DIR"); v != "" {
+		credsDir = v
+	}
+	return findCredentialsInDir(repoURL, credsDir)
+}
+
+// findCredentialsInDir searches for credentials for repoURL in configuration
+// files in credsDir directory. If the directory does not exist, empty
+// credentials structure with nil err is returned.
+// Only files that do not begin with dot and have either no or ".conf"
+// extension are searched. The files are searched in ascending lexicographic
+// order. The first file that contains machine declaration matching repoURL
+// ends the search. If no file contain matching machine declaration, empty
+// credentials structure with nil err is returned.
+func findCredentialsInDir(repoURL string, credsDir string) (creds credentials, err error) {
+	contents, err := os.ReadDir(credsDir)
+	if err != nil {
+		if os.IsNotExist(err) {
+			err = nil
+			debugf("credentials directory %#v does not exist", credsDir)
+		} else {
+			err = fmt.Errorf("cannot open credentials directory: %w", err)
+		}
+		return
+	}
+
+	creds, query, err := parseRepoURL(repoURL)
+	if err != nil {
+		err = fmt.Errorf("cannot parse archive URL: %w", err)
+		return
+	}
+	if query == nil { // creds.Empty() == false
```

Comment:
This needs to be improved a bit so that the intention is more clear.  If the point here is that we already have credentials, we should check creds explicitly instead of checking query and assuming credentials are set in a comment.

Reply:
Replaced with `!creds.Empty()`. I can't rely on `err != nil` here as that alone does not tell me whether the URL contained credentials. @niemeyer Does it resolve this thread?